lean/bitvec.lean

```diff
@@ -0,0 +1,375 @@
+    | 0 _  _:= 0
+    | (nat.succ w) x y :=
+      let xr := x.bodd_div2 in
+      let yr := y.bodd_div2 in
+      nat.bit (f xr.fst yr.fst) (fin_bitwise w xr.snd yr.snd)
+
+  -- A theorem that nat.bit is less than a power of two, when the input
+  -- is.
+  --
+  -- The implicit parameters are chosen so that apply is useful here.
+  theorem bit_lt_pow2  {w:ℕ} {b : bool} {x : ℕ} (h : x < 2^w)
+  : nat.bit b x < 2^(w+1) :=
+  begin
+    -- Simplify 2^(w+1)
+    simp [nat.add_succ, nat.pow_succ, nat.mul_comm _ 2, eq.symm (nat.bit0_val _)],
+    -- Split on b and simplify bit
+    cases b; simp [bit],
+    case ff { apply nat.bit0_lt h, },
+    case tt { apply nat.bit1_lt_bit0 h, }
+  end
+
+  theorem bitwise_lt (f: bool → bool → bool) (w:ℕ)
+  : ∀(x y : ℕ),  fin_bitwise f w x y < 2^w :=
+  begin
+    induction w with w p,
+    { intros, dec_trivial_tac, },
+    intros,
+    unfold fin_bitwise,
+    apply bit_lt_pow2 (p _ _),
+  end
+
+  def bitwise {w:ℕ} (f: bool → bool → bool) (x y : bitvec w) : bitvec w :=
+    ⟨fin_bitwise f w x.val y.val, bitwise_lt _ _ _ _⟩
+
+  def not {w:ℕ} (x: bitvec w) : bitvec w := ⟨2^w - x.val - 1,
+    begin
+      have xval_pos : 0 < x.val + 1,
+      { apply (succ_pos x.val) },
+      apply (sub_lt _ xval_pos),
+      apply pos_pow_of_pos,
+      apply (succ_pos 1)
+    end⟩
+
+  def and {w:ℕ} : bitvec w → bitvec w → bitvec w := bitwise band
+  def or  {w:ℕ} : bitvec w → bitvec w → bitvec w := bitwise bor
+  def xor {w:ℕ} : bitvec w → bitvec w → bitvec w := bitwise bxor
+
+  infix `.&&.`:70 := and
+  infix `.||.`:65 := or
+
+end bitwise
+
+
+section shift
+  -- Shift related operations, including signed and unsigned shift.
+
+  variable {n : ℕ}
+
+  -- shift left
+  def shl (x : bitvec n) (i : ℕ) : bitvec n := ⟨x.val * 2^i % 2^n, by simp⟩
+
+  -- unsigned shift right
+  def ushr (x : bitvec n) (i : ℕ) : bitvec n := ⟨x.val / 2^i, by exact div_pow_mono x.property⟩
+
+  section listlike
+    -- This is really a listlike operation, but we define it early so
+    -- so that we can use it to define sshr.
+    def msb {n:ℕ} (x: bitvec n) : bool := (ushr x (n-1)).val = 1
+  end listlike
+
+  -- signed shift right
+  def sshr (x: bitvec n) (i:ℕ) : bitvec n :=
+    -- When the sign bit is set in x, (msb x = 1), then we would like
+    -- the result of sshr x i, to have the top i bits set.
+    -- We can calculate a number that will do this in steps:
+    -- 1) (2 ^ n) - 1, will have all the bits set.
+    -- 2) (2 ^ (n-i)) - 1, will be a number with the lower (n-i) bits set.
+    -- 3) subtract the previous two numbers (1) - (2), to get a
+    -- number where only the top i bits are set.
+    let upper_bits := 2 ^ n - 2 ^ (n-i) in
+    let sign := if msb x then upper_bits else 0 in
+    ⟨ sign + (ushr x i).val,
+      begin
+        simp [sign], cases (msb x),
+        case bool.ff
+        { simp [ushr], apply div_pow_mono x.property },
+        case bool.tt
+        { apply sshr_in_range x.property }
+      end⟩
```

Comment:
This could provably be implemented by defining `def {n} (x:bitvec n) : \int` that maps a bitvector to its signed representation, then using int.shiftr, and mapping back to a bitvector.